design/machine-remediation-request.md

```diff
@@ -0,0 +1,142 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# machine-remediation-request
+
+## Status
+
+in-progress
+
+## Table of Contents
+
+<!--ts-->
+* [machine-remediation-request](#machine-remediation-request)
+  * [Status](#status)
+  * [Table of Contents](#table-of-contents)
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+  * [Proposal](#proposal)
+    * [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+      * [MachineRemediationRequest spec API](#machineremediationrequest-spec-api)
+      * [MachineRemediationRequest status API](#machineremediationrequest-status-api)
+    * [Risks and Mitigations](#risks-and-mitigations)
+  * [Design Details](#design-details)
+    * [Work Items](#work-items)
+    * [Dependencies](#dependencies)
+    * [Test Plan](#test-plan)
+    * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+    * [Version Skew Strategy](#version-skew-strategy)
+  * [Alternatives](#alternatives)
+  * [References](#references)
+
+<!-- Added by: dhellmann, at: Wed Jun 26 16:08:57 EDT 2019 -->
+
+<!--te-->
+
+## Summary
+
+This document explains the implementation of the Machine Remediation Request controller and
+how it will work together with Machine Health Check controller to restore unhealthy machines.
+
+## Motivation
+
+To provide a mechanism to apply remediation strategy on an unhealthy node, that will be different by type and by a provider.
+
+### Goals
+
+1. To provide the interface that will make possible to apply different remediation strategy for each cloud or bare metal providers.
```

Comment:
The way I see it, "remediation" is a machine health check concept, and the machine health check controller should be the one responsible for the remediation strategy. It seems very strange to me that each machine api provider would figure out its own remediation strategy.

The machine API should provide low-level actions for implementing a remediation strategy, and perhaps per-machine metadata to help it make better remediation choices.

i.e. something like

```
machinehealthcheck_controller:

  remediate() {
      if !machine.hasAnnotation("doNotRemediateWithReboot") && machine->reboot  != null {
          machine->reboot()
      } else {
          machine->delete()
      }   
  }
```

i.e. perhaps not all provider will implement reboot, and perhaps we can allow some indication to not try reboot even if the provider supports it ... but otherwise we should try reboot first

(Sorry for the pseudo-code - I know it omits a lot of complications about tracking progress and falling back to delete if reboot doesn't work)

But IMO ... the conclusion for metal3 is that the Machine API should have a reboot interface

Reply:
IMHO the one who is implementing actuator should be more aware regarding the needed action for the remediation, MHC should only monitor node conditions and apply the remediate logic in the case when it needed, but it should not be aware regarding the implementation of the remediation logic.
It similar to the cluster API and actuators relations, cluster API only applies that actuator logic.

IIRC the community has some proposals to introduce the reboot method and the actuator interface under the actuator interface, but without luck, because reboot and re-provision for the cloud provider is quite the same, also after the interface change you will need to change all controllers that use this interface

Reply:
We're in basic agreement, the challenge has been that upstream has been slow moving (to the point that OpenShift forked it) and primarily focused on cloud use cases which are less likely to face the challenges of a hyper-converged bare metal environment.

While there may be more cause for optimism now under VMware's leadership, getting the API changes through and backported in time for OpenShift feature freeze is unrealistic IMHO. Originating changes in the OpenShift Machine API fork first is also not an option.

Hence the idea to have $SomethingElse that fills the same role in the short term, can be used to demonstrate value in upstream conversations, and deprecated in the longer term.

On the subject though, I see value in adding the following to Machine API:
- reboot()
- on()
- off(), if rebooting the node doesn't result in it becoming healthy again, then we'd want to turn it off and move on
- remediate(), an opportunity to hide even the details of how to return the node to health from MHC.